sympy/solvers/solveset.py

```diff
@@ -2514,6 +2596,11 @@ def linear_eq_to_matrix(equations, *symbols):
     The only simplification performed is to convert
     ``Eq(a, b)`` $\Rightarrow a - b$.
 
+    If the system is known to be linear and is in an expanded
+    state, the flag `literal=True` can be used; this will also
+    allow non-symbols to appear in ``symbols``. There is little
+    error checking for non-linearity, so use with caution.
```

Comment:
Some example is needed in the docstring. I don't think that the flag name "literal" makes much sense here. At least to me I don't have a sense for what either `literal=True` or `literal=False` mean.

I still wonder if we can't just make `linear_eq_to_matrix` handle the cases of interest by default though. There is not necessarily anything conceptually wrong with writing nonlinear equations in terms of matrices e.g.:
```python
In [5]: linear_eq_to_matrix([z + x + y, x - y], [x, y])
Out[5]: 
⎛⎡1  1 ⎤  ⎡-z⎤⎞
⎜⎢     ⎥, ⎢  ⎥⎟
⎝⎣1  -1⎦  ⎣0 ⎦⎠

In [6]: linear_eq_to_matrix([x**2 + x + y, x - y], [x, y])
---------------------------------------------------------------------------
NonlinearError
```
If the returned result had included `x**2` it would still satisfy the key invariant:
```python
In [7]: A, b = Matrix([[1, 1], [1, -1]]), Matrix([-x**2, 0])

In [8]: A, b
Out[8]: 
⎛         ⎡  2⎤⎞
⎜⎡1  1 ⎤  ⎢-x ⎥⎟
⎜⎢     ⎥, ⎢   ⎥⎟
⎝⎣1  -1⎦  ⎣ 0 ⎦⎠

In [9]: A*Matrix([x, y]) - b
Out[9]: 
⎡ 2        ⎤
⎢x  + x + y⎥
⎢          ⎥
⎣  x - y   ⎦
```
Likewise `linsolve` still gives a valid result if correctly interpretted:
```python
In [10]: linsolve((A, b))
Out[10]: 
⎧⎛  2     2 ⎞⎫
⎪⎜-x    -x  ⎟⎪
⎨⎜────, ────⎟⎬
⎪⎝ 2     2  ⎠⎪
⎩            ⎭
```
This implies that the solutions are given by `x = -x**2/2` and `y = x`:
```python
In [12]: solve([x**2 + x + y, x - y], [x, y], dict=True)
Out[12]: [{x: -2, y: -2}, {x: 0, y: 0}]

In [13]: solve([Eq(x, -x**2/2), Eq(y, -x**2/2)], [x, y], dict=True)
Out[13]: [{x: -2, y: -2}, {x: 0, y: 0}]
```
Of course in some contexts you would want `linear_eq_to_matrix` and `linsolve` to raise for these kinds of inputs and there should be some way to ask for that behaviour (e.g. a `strict=True` flag or something).